relay-general/src/protocol/replay.rs

```diff
@@ -0,0 +1,567 @@
+            .contexts
+            .value_mut()
+            .as_mut()
+            .unwrap()
+            .get("browser")
+            .unwrap()
+            .clone();
+
+        let loaded_os_context = replay_value
+            .contexts
+            .value_mut()
+            .as_mut()
+            .unwrap()
+            .get("client_os")
+            .unwrap()
+            .clone();
+
+        let loaded_device_context = replay_value
+            .contexts
+            .value_mut()
+            .as_mut()
+            .unwrap()
+            .get("device")
+            .unwrap()
+            .clone();
+
+        assert_eq!(loaded_browser_context, browser_context);
+        assert_eq!(loaded_os_context, os_context);
+        assert_eq!(loaded_device_context, device_context);
+    }
+
+    #[test]
+    fn test_missing_user() {
+        let payload = include_str!("../../tests/fixtures/replays/replay_missing_user.json");
+
+        let mut replay: Annotated<Replay> = Annotated::from_json(payload).unwrap();
+        let replay_value = replay.value_mut().as_mut().unwrap();
+        replay_value.normalize(None, None);
+
+        let user = replay_value.user.value();
+        assert!(user.is_none());
+    }
+
+    #[test]
+    fn test_set_ip_address_missing_user_ip_address() {
+        let ip_address = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));
+
+        // IP-Address set.
+        let payload =
+            include_str!("../../tests/fixtures/replays/replay_missing_user_ip_address.json");
+
+        let mut replay: Annotated<Replay> = Annotated::from_json(payload).unwrap();
+        let replay_value = replay.value_mut().as_mut().unwrap();
+        replay_value.normalize(Some(ip_address), None);
+
+        let ipaddr = replay_value
+            .user
+            .value_mut()
+            .as_ref()
+            .unwrap()
+            .ip_address
+            .value()
+            .unwrap()
+            .as_str();
+        assert!("127.0.0.1" == ipaddr);
+    }
+
+    #[test]
+    fn test_loose_type_requirements() {
+        let payload = include_str!("../../tests/fixtures/replays/replay_failure_22_08_31.json");
+
+        let mut replay: Annotated<Replay> = Annotated::from_json(payload).unwrap();
+        let replay_value = replay.value_mut().as_mut().unwrap();
+        replay_value.normalize(None, None);
+
+        let user = replay_value.user.value_mut().as_mut().unwrap();
+        assert!(user.ip_address.value_mut().as_mut().unwrap().as_str() == "127.1.1.1");
+        assert!(user.username.value_mut().is_none());
+        assert!(user.email.value_mut().as_mut().unwrap().as_str() == "email@sentry.io");
+        assert!(user.id.value_mut().as_mut().unwrap().as_str() == "1");
+    }
+
+    #[test]
+    fn test_scrub_pii_from_annotated_replay() {
+        let mut pii_config = PiiConfig::default();
+        pii_config.applications =
+            BTreeMap::from([(SelectorSpec::And(vec![]), vec!["@common".to_string()])]);
```

Comment:
When we fixed this test, I did not realize you were planning to read PII config from project configs instead of using a hard-coded one. So it might make sense to change this test back to a pii config which looks like the most common datascrubbing settings coming from sentry:

https://github.com/getsentry/relay/blob/e87b92e261ffe4f563c8e5519c12c7c1e756984d/relay-general/src/pii/convert.rs#L196-L203

But with that, even the credit card scrubbing seems to fail and I'm not sure why. I can investigate later this week.